File: dynamic-programming/grid/triangle/ts/solution_dfs.ts
const NOT_CALCULATED_VALUE = undefined

function minimumTotal(triangle: number[][]): number {
    const n = triangle.length
    let start = 0
    const memory: number[][] = Array.from({length: n}, (): number[] => Array(++start).fill(NOT_CALCULATED_VALUE))
    const dfs = (i: number, j: number): number => {
        if (i === n - 1) {
            return triangle[i][j]
        }
        if (memory[i][j] !== NOT_CALCULATED_VALUE) {
            return memory[i][j]
        }
        return memory[i][j] = Math.min(dfs(i + 1, j), dfs(i + 1, j + 1)) + triangle[i][j]
    }
    return dfs(0, 0)
}

export { minimumTotal }

File: dynamic-programming/grid/triangle/java/Solution_Dfs.java
import java.util.Arrays;
import java.util.List;
import java.util.function.BiFunction;

class Solution_Dfs {
    private final static Integer NOT_CALCULATED_VALUE = null;

    public int minimumTotal(List<List<Integer>> triangle) {
        int n = triangle.size();
        Integer[][] memory = new Integer[n][n];
        for (Integer[] row : memory) {
            Arrays.fill(row, NOT_CALCULATED_VALUE);
        }
        BiFunction<Integer, Integer, Integer> dfs = new BiFunction<Integer, Integer, Integer>() {
            @Override
            public Integer apply(Integer i, Integer j) {
                if (i == n - 1) {
                    return triangle.get(n - 1).get(j);
                }
                if (memory[i][j] != NOT_CALCULATED_VALUE) {
                    return memory[i][j];
                }
                return memory[i][j] = Math.min(this.apply(i + 1, j), this.apply(i + 1, j + 1))
                        + triangle.get(i).get(j);
            }
        };
        return dfs.apply(0, 0);
    }
}